src/ot/matching.rs

```diff
@@ -111,61 +109,71 @@ pub fn match_input(
                 }
 
                 if ligbase == Ligbase::MayNotSkip {
-                    return None;
+                    return false;
                 }
             }
         } else {
             // If first component was NOT attached to a previous ligature component,
             // all subsequent components should also NOT be attached to any ligature
             // component, unless they are attached to the first component itself!
             if this_lig_id != 0 && this_lig_comp != 0 && (this_lig_id != first_lig_id) {
-                return None;
+                return false;
             }
         }
 
         total_component_count += this.lig_num_comps();
     }
 
-    Some(Matched {
-        len: iter.index() - ctx.buffer.idx + 1,
-        positions,
-        total_component_count,
-    })
+    *end_position = iter.index() + 1;
+
+    if let Some(p_total_component_count) = p_total_component_count {
+        *p_total_component_count = total_component_count;
+    }
+
+    return true;
 }
 
 pub fn match_backtrack(
-    ctx: &ApplyContext,
+    ctx: &mut ApplyContext,
     backtrack_len: u16,
     match_func: &MatchingFunc,
-) -> Option<usize> {
+    match_start: &mut usize,
+) -> bool {
     let mut iter = SkippyIter::new(ctx, ctx.buffer.backtrack_len(), backtrack_len, true);
     iter.enable_matching(match_func);
 
     for _ in 0..backtrack_len {
-        if !iter.prev() {
-            return None;
+        let mut unsafe_from = 0;
+        if !iter.prev(Some(&mut unsafe_from)) {
```

Comment:
Jesus... Can you point how it looks in harfbuzz?

Reply:
Here:

https://github.com/harfbuzz/harfbuzz/blob/ac46c3248e8b0316235943175c4d4a11c24dd4a9/src/hb-ot-layout-gsubgpos.hh#L1205-L1230

That pattern appears a couple of times unfortunately (i.e. having arguments with default values), which doesn't exist in Rust... So in order to stay as close to the original as possible, I decided to use options instead and unwrap them with the default value in the method. Which results in a lot of clutter, but I think that's as close as it gets to the original code.